mesop/web/src/services/channel.ts

```diff
@@ -117,143 +139,204 @@ export class Channel {
-                    `Number of diffs (${numDiffStates}) doesn't equal the number of states (${numStates}))`,
-                  );
-                }
-
-                // `this.states` should be populated at this point since the first update
-                // from the server should be the full state.
-                for (let i = 0; i < numDiffStates; ++i) {
-                  const state = applyStateDiff(
-                    this.states.getStatesList()[i].getData() as string,
-                    states.getStatesList()[i].getData() as string,
-                  );
-                  this.states.getStatesList()[i].setData(state);
-                }
-                break;
-              }
-              case UpdateStateEvent.TypeCase.TYPE_NOT_SET:
-                throw new Error('No state event data set');
-            }
+        console.debug('Server event (SSE): ', uiResponse.toObject());
+        this.handleUiResponse(request, uiResponse, initParams);
+      });
+    });
+  }
+
+  private initWebSocket(initParams: InitParams, request: UiRequest) {
+    if (this.socket) {
+      this.status = ChannelStatus.OPEN;
+      const payload = generatePayloadString(request);
+      this.socket.emit('message', payload);
+      return;
+    }
+    this.socket = io('/__ui__', {
+      transports: ['websocket'],
+      reconnectionAttempts: 3,
+    });
+
+    this.status = ChannelStatus.OPEN;
+    this.isWaitingTimeout = setTimeout(() => {
+      this.isWaiting = true;
+    }, WAIT_TIMEOUT_MS);
+
+    this.logger.log({type: 'StreamStart'});
+
+    const {zone, onRender, onError, onCommand} = initParams;
+    this.initParams = initParams;
+
+    this.socket.on('connect', () => {
+      // Send the initial UiRequest upon connection
+      const payload = generatePayloadString(request);
+      this.socket!.emit('message', payload);
+    });
+
+    this.socket.on('response', (data: any) => {
+      const prefix = 'data: ';
+      const payloadData = (data.data.slice(prefix.length) as string).trimEnd();
+      zone.run(() => {
+        if (payloadData === STREAM_END) {
+          this._isHotReloading = false;
+          this.status = ChannelStatus.CLOSED;
+          this.logger.log({type: 'StreamEnd'});
+          this.dequeueEvent();
+          return;
+        }
+
+        const array = toUint8Array(atob(payloadData));
+        const uiResponse = UiResponse.deserializeBinary(array);
+        console.debug('Server event (WebSocket): ', uiResponse.toObject());
+        this.handleUiResponse(request, uiResponse, initParams);
+      });
+    });
+
+    this.socket.on('error', (error: any) => {
+      zone.run(() => {
+        this.socket = undefined;
+        console.error('WebSocket error:', error);
+        this.status = ChannelStatus.CLOSED;
+      });
+    });
+
+    this.socket.on('disconnect', (reason: string) => {
+      zone.run(() => {
+        this.socket = undefined;
+        console.error('WebSocket disconnected:', reason);
```

Comment:
I think fine for now, but could be good to notify the user if the socket gets into a disconnected or error state.

Reply:
Good point - filed #1033 - definitely should improve this, perhaps some kind of auto-reconnect 3 times and then show error message.